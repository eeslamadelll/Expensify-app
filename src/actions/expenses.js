import database from '../firebase/firebase';

// creating an action generator for the add expense
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {description = '', note = '', amount = 0, createdAt = 0} = expenseData;
        const expense = {description, note, amount, createdAt};
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        }).catch((e) => {
            console.log('Error: ', e);
        });
    };
};

// creating an action generator for the remove expense
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({id} = {}) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove(() => {
            dispatch(removeExpense({ id }));
        });
    }
}

// creating an action generator for the edit expense
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

// creating an action generator for the set expenses
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// creating the async startSetExpenses action to fetch from firebase DB
export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        });
    }
}