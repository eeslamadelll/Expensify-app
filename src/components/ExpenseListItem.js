import React from 'react';
import {Link} from 'react-router-dom';


const ExpenseListItem = ({id, description, amount, createdAt}) => {
    return (
        <div>
            <Link to={`/edit/${id}`}>
                <p>Description : { description }</p>
            </Link>
            <span>Created At : { createdAt }</span>
            <p>Amount : { amount }</p>
        </div>
    )
};

export default ExpenseListItem;