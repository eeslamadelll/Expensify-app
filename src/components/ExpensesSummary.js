import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';


export const ExpensesSummary = ({expenseCount, expenseTotal}) => {
    const expenseWord = expenseCount > 1 ? 'expenses' : 'expense';
    return(
        <div className="page-header">
            <div className="content-container">
                <p>you have <span>{expenseCount}</span> {expenseWord} with total of <span>{numeral(expenseTotal / 100).format('$0,0.00')}</span></p>
                <Link className="add" to="/add">Add Expense</Link>
            </div>
        </div>
    );
}

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expenseTotal: getExpensesTotal(visibleExpenses)
    }
};

export default connect(mapStateToProps)(ExpensesSummary);