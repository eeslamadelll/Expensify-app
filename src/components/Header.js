import React from 'react';
import {NavLink} from 'react-router-dom';

const Header = () => {
    return (
        <header>
            <h1>Expensify</h1>
            <nav>
                <ul>
                    <li><NavLink to="/" activeClassName="is-active" exact={true}>Homepage</NavLink></li>
                    <li><NavLink to="/add" activeClassName="is-active">Add Expense</NavLink></li>
                </ul>
            </nav>
        </header>
    );
};

export default Header;