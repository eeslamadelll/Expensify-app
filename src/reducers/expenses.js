// creating the expenses reducer
const defaultExpenseState = [];
export default (state = defaultExpenseState, action) => {
    switch(action.type){
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            return state.filter( (expense) => expense.id !== action.id );
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if(expense.id === action.id){
                    return{
                        ...expense,
                        ...action.updates
                    };
                }else{
                    return expense;
                }
            });
        default:
            return state;
    }
}
