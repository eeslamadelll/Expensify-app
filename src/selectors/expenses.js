import moment from 'moment';

// get visible expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true; // if start date is the same or before the creation of the expense
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true; // if end date is the same or after the creation of the expense
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        return textMatch && startDateMatch && endDateMatch;
    }).sort((a, b) => {
        if(sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1 : -1;
        }else if(sortBy === 'amount'){
            return a.amount < b.amount ? 1 : -1;
        }
    });
}