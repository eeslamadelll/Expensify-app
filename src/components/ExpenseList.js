import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
    	{
    		props.expenses.length === 0 ? (
    			<p>There is no expenses to show</p>
    		) : (
    			props.expenses.map((expense) => {
			        return <ExpenseListItem key={expense.id} {...expense}/>;
			    })
    		)
    	}
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList);


///////////////////// this code is right but we will use the best practice
// const ConnectExpenseList = connect((state) => {
//     return {
//         expenses: state.expenses,
//         filters: state.filters
//     };
// })(ExpenseList);

// export default ConnectExpenseList;