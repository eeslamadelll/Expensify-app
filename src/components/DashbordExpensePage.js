import React from 'react';
import ExpenseList from './ExpenseList'; // data comes from HOC
import ExpenseListFilters from './ExpenseListFilters';

const DashbordExpensePage = () => {
    return (
        <div>
            <ExpenseListFilters />
            <ExpenseList />
        </div>
    );
}

export default DashbordExpensePage;