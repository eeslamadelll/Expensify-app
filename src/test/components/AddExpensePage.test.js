import React from 'react';
import { shallow } from 'enzyme';
import { AddExpensePage } from '../../components/AddExpensePage';
import expenses from '../fixture/expenses';

test('Should Render AddExpensePage correctly', () => {
    const addExpense = jest.fn();
    const history = { push: jest.fn() };
    const wrapper = shallow(<AddExpensePage addExpense={addExpense} history={history} />)
    expect(wrapper).toMatchSnapshot();
});

test('Should handle on submit', () => {
    const addExpense = jest.fn();
    const history = { push: jest.fn() };
    const wrapper = shallow(<AddExpensePage addExpense={addExpense} history={history} />)
    wrapper.find('AddExpenseForm').prop('onSubmit')(expenses[0]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(addExpense).toHaveBeenLastCalledWith(expenses[0]);
});