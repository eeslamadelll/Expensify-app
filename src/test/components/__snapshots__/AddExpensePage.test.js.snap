// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Render AddExpensePage correctly 1`] = `
<div>
  <h1>
    Add Expense
  </h1>
  <AddExpenseForm
    onSubmit={[Function]}
  />
</div>
`;
